### DISCUSSION TITLE
TCL/Tk wrapper for zxing-cpp

### DISCUSSION THREAD

**Discussion by oehhar** [zxing-cpp/zxing-cpp#876] [Show and tell]
Hi Axel, hi great zxing-cpp team,
I started a proof of concept of a [zxing-cpp wrapper for Tcl/Tk](https://github.com/oehhar/zxing-cpp-wrapper-tcl): 
An image may be given as raw binary or an image object and decoded.
If there are TkInter folks here - this is a native extension to decode TkInter photo images.

My aim is to replace zbar in [undroidwish](https://www.androwish.org) and to do a webcam capture and decode like in [zbartool.tcl](https://www.androwish.org/home/file?name=assets/zbartool0/zbartool.tcl&ci=tip).
This is cross-platform including Android, MacOS, Unix and Windows.

The aim is to copy all features of [tclzbar.c](https://www.androwish.org/home/file?name=jni/ZBar/tcl/tclzbar.c&ci=tip) by magic Christian Werner.

The next open steps are:

- Set zxing-cpp options
- background processing in a worker thread
- Unix/MacOS make file
- Do a pull request to propose it to the main repository

TCL uses a configure or nmake toolchain. Only the nmake toolchain is tested with VisualStudio 2022.

Thanks for any comment and take care,
Harald

**Comment by axxel** [comment]
Thanks for sharing. May I ask what your motivation is? No offense, but I thought TCL/TK was a somewhat obsolete technology?

**Comment by oehhar** [comment]
Thanks for your opinion.
I presented the tcl/tk wrapper at the Vienna OpenACS/Tcl/Tk conference this summer. [First speech on Friday](https://openacs.km.at/evaluate/org/129998253/companyhelp/schedule)

It is correct, that Tcl/Tk is relatively unvisible.
One may use a more visible project like TkInter Python. A Demo might:

- get constantly images from the Webcam (see androwish.org how to do this on Android/Linux/MacOS/Windows)
- Decode with zxing-cpp and only show images, where decodes are tried
- Foreward this to Python

Remark, that python TkInter is written in TCL/Tk.
One may do the same with Perl/Tk or Ruby/Tk

Version 1.1 of the code is now released adding the "TryDenoise" option and merge current main branch.
I use it in productive environments now for one year and I am very satisfied.

The points of my initial post, which are now fullfilled are:

   * Set zxing-cpp options
   * background processing in a worker thread

The Unix/MacOS make files are not ready.
Androwish/Undroidwish does not want to support it, as the C++ requirements are quite high and cmake does not fit to the build system.

Thanks for great ZXING-CPP !

Harald

**Comment by oehhar** [comment]
Another point. I don't find the original ticket. I have made a lot of tests with ECI and GS1 codes. The output "bytesECI" now behaves correctly as an ISO/IEC compliant decoder.
Thanks for all, great work!
Harald
